File: LeetCode/List&LinkList/283.move-zeroes.py
#
# @lc app=leetcode id=283 lang=python3
#
# [283] Move Zeroes
#

# @lc code=start
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        pEdit = 0     
        pFilter = 0   
        while(pFilter < len(nums)):
            if(nums[pFilter] != 0):
                nums[pEdit] = nums[pFilter]
                pEdit+=1
            pFilter+=1
        for i in range(pEdit, len(nums)):
            nums[i] = 0

# pEdit应该永远指向第一个零的位置 
# pFilter寻找非0元素当找到一个非0元素它就与pEdit交换， 并且把pEdit指向下一个0
# @lc code=end

# 010312
# 100312
# 130012
# 131002
# 131200

File: LeetCode_Go/List&LinkedList/283.move-zeroes.go
/*
 * @lc app=leetcode id=283 lang=golang
 *
 * [283] Move Zeroes
 */

// @lc code=start

// Given an integer array nums, move all 0's to the end of it  in-place, while maintaining the relative order of the non-zero elements.

// two pointers: `curIdx`, for populate the `non-zero slice` from the beginning of the array; `searchIdx`, for traverse the array and find non-zero element. Each time a non-zero element is found assign it to the curIdx, and move curIdx to the right. While the bound of non-zero slice is found, fill the rest with 0.

func moveZeroes(nums []int) {
	curIdx, searchIdx := 0, 0
	for searchIdx < len(nums) {

		// fmt.Println(nums)
		// fmt.Printf("searchIdx: %d, curIdx: %d \n", searchIdx, curIdx)

		if nums[searchIdx] != 0 {
			nums[curIdx] = nums[searchIdx]
			curIdx++
		}

		searchIdx++
	}

	for i := curIdx; i < len(nums); i++ {
		nums[i] = 0
	}
}

// @lc code=end

